library/os_container_infra.py

```diff
@@ -0,0 +1,136 @@
+        elif container_infra_details.status.endswith('PROGRESS'):
+            # If the cluster creation and update is in progress:
+            display.warning(container_infra_details.status)
+            return False
+        elif container_infra_details.status.endswith('COMPLETE'):
+            # If the cluster looks ready
+            # Raise error if there is a mismatch between cluster template ids
+            cluster_template_mismatch = kwargs['cluster_template_id'] != container_infra_details.cluster_template_id 
+            if cluster_template_mismatch:
+                module.fail_json(
+                    msg="Cluster template mismatch for an existing cluster.")
+
+            patch_list = list()
+
+            # Update number of masters if there is a mismatch
+            master_count_mismatch = kwargs['master_count'] != container_infra_details.master_count
+            if master_count_mismatch:
+                patch = dict(op='replace',
+                     value=kwargs['master_count'],
+                     path='/master_count')
+                patch_list.append(patch)
+
+            # Update number of nodes if there is a mismatch
+            node_count_mismatch = kwargs['node_count'] != container_infra_details.node_count
+            if node_count_mismatch:
+                patch = dict(op='replace',
+                     value=kwargs['node_count'],
+                     path='/node_count')
+                patch_list.append(patch)
+
+            if patch_list:
+                # There appears to be a patch needed to apply
+                magnum_client.clusters.update(container_infra_details.uuid, patch_list)
+                display.warning("Applying patch to the cluster: {}".format(patch_list))
+                return False
+            else:
+                # Everything looks okay, return cluster details
+                return container_infra_details
+        else:
+            module.fail_json(
+                msg="Cluster returned unexpected status: {}".format(container_infra_details.status))
+    except NotFound:
+        # Create new cluster here if no cluster with the name was found
+        magnum_client.clusters.create(**kwargs)
+        display.warning("Cluster called {} does not exist, creating new.".format(kwargs['name']))
+        return False
+
+def main():
+    module = AnsibleModule(
+        argument_spec = dict(
+            cloud_name=dict(required=True, type='str'),
+            magnum_endpoint=dict(required=True, type='str'),
+            cluster_name=dict(required=True, type='str'),
+            cluster_template_name=dict(required=True, type='str'),
+            master_count=dict(required=True, type='int'),
+            node_count=dict(required=True, type='int'),
+            keypair=dict(required=True, type='str'),
+        ),
+        supports_check_mode=False
+    )
+
+    cloud_name = module.params['cloud_name']
+    magnum_endpoint = module.params['magnum_endpoint']
```

Comment:
This shouldn't be required - we should be able to use the keystone endpoint catalogue to look up the magnum endpoint. Does it not work without endpoint_override?